updated package hub drawer (#685)
SimpleDataLabsInc/prophecy-docs: docs/extensibility/package-hub/build-a-package.md

Importantly, access to the actual data depends on your personal access token or username/password credentials. To access the data in a new project, you must select a fabric with access permissions for that data.

## Pipelines

You can share pipelines from one project to another project and run it with a [Config](/engineers/configurations) in the dependent projects. This allows data administrators to create deployment templates for the pipelines that use the best practices for authorization, notifications, error handling, and logging information.

:::note
If you modify anything from an existing or new pipeline, you must publish these updates as a new version for the package to pick up the changes. Then, you must manually update the project dependency to get the updated package version.
:::

### Prepare a pipeline

To update a Config for a pipeline:

1. Open the pipeline you want to create a Config for.

1. At the top of the window, click **Config**.

1. Click the **Config** tab.

1. On the top right, select the Config you want to update.

> 1. Update the values in teh Config.

Reviewer: ```suggestion
1. Update the values in the Config.
```

Reviewer: Actually, I believe the correct steps are to create a config schema. This is the set of variables that end users will assign values to. I think you should review how to use pipeline configurations, and test it too. https://docs.prophecy.io/engineers/configurations

<img width="1054" alt="image" src="https://github.com/user-attachments/assets/ac365080-b6bd-4f1e-a61c-d0074d9504dd" />

1. Click **Save**.

### Interactively run a pipeline

If you interactively run a pipeline from your current project, you need to create a Config in your other project. Then, the pipeline is visible in other project.

To interactively run a pipeline:

1. Hover over the play button at the bottom right corner of the screen, and click **...**.

   The **Interactive Run Configuration** modal appears.

1. (Optional) If you want to only use a subset of your data for the input, toggle on **Limit Input records** and enter the amount you want to use.

1. (Optional) If you want to no longer use the pipelines data sampling settings, toggle off **Data Sampling**.

   By default, data sampling is enabled. When left enabled, Data sampling uses the pipeline's data sampling settings. Prophecy samples data during the interactive run experience to provide the best debugging experience for users.

1. Select the configuration you want to run in the **Configuration** drop down.

1. Click the play button to run your pipeline.

### Schedule a pipeline

> To schedule a pipeline to run:

Reviewer: ```suggestion
To schedule a pipeline that you have imported from a package:
```

1. In the left side bar, hover over **Jobs** and click **+**.

1. Enter your jobs **Name**, and update the **Schedule Interval** to how often you would like it to run.

   To learn more about scheduling, see [Scheduling](/analysts/scheduling).

1. Assure the other fields are accurate.

1. Click **Create New**.

>    Notice that your new schedule appears under **Jobs** in teh left side bar.

Reviewer: ```suggestion
   Notice that your new schedule appears under **Jobs** in the left side bar.
```

1. At the top left of the screen, click **Pipeline**.

> 1. Open the Pipeline gem, and select the pipeline you want to schedule.

Reviewer: ```suggestion
1. Open the Pipeline gem, and select the pipeline you want to schedule. All pipelines in your project, including any pipelines from packages, should appear here .
```

> 1. Select the configuration you want to run with the pipeline.

Reviewer: Okay I need this to be clarified. Can you create a new pipeline config for the imported pipeline inside the project that is NOT the original project? Or do all configs need to be set up in the original project? Or can it be both?

---
title: Build a Package
id: build-a-package
slug: /engineers/build-a-package
description: Learn how to build a package in Package Hub where you can share datasets, pipelines, subgraphs, and UDFs
tags: []
---

Within Package Hub, you can build your package where you can share [datasets](#datasets), [pipelines](#pipelines), [subgraphs](#subgraphs), and [user-defined functions](#user-defined-functions) across projects.

## Datasets

> You can share a [dataset](/engineers/dataset) across your pipelines, but be cautious when you share it across projects.

Author: (For Monday) Question: So, why exactly do we want to share datasets in packages? And what are the limitations?

> You can share a [dataset](/engineers/dataset) across your pipelines, but be cautious when you share it across projects.

Reviewer: Okay I think it is:

- Share datasets in packages so others can use your set up like custom schema, properties, etc.
- Limitations = if a dataset requires certain authentication that you don't have, then it won't work for you

Can you confirm with an SME? Maciej maybe? Or post in #product-help?

### Datasets across pipelines

When you add a new dataset as a Source or Target in your pipeline, you can immediately use the dataset in all your pipelines in that project. You can also use that dataset as a Source or Target in other pipelines.

If you modify the dataset in one pipeline, Prophecy automatically modifies the dataset in each pipeline using it.

### Datasets across projects

A dataset in Prophecy points to your actual data in your data storage solution. Prophecy does not store any data, so we recommend not configuring datasets in a package and leaving datasets configurable in the project where you use them.

Importantly, access to the actual data depends on your personal access token or username/password credentials. To access the data in a new project, you must select a fabric with access permissions for that data.

## Pipelines

You can share pipelines from one project to another project and run it with a Config in the dependent projects. This allows data administrators to create deployment templates for the pipelines that use the best practices for authorization, notifications, error handling, and logging information.

> ### Example scenario

Reviewer: Yes, because it looks like configs are a prereq for sharing pipelines in a useable way. The video is waaaay old, use this doc instead: https://docs.prophecy.io/engineers/configurations

This is essentially how users share pipelines for Spark without having business apps like we have for SQL

For example, if your data administrator creates a pipeline in a project called `base_project`, they can add Config variables to the pipeline. Then, if you reference `base_project` in another project called `new_project`, you can reference the Config variables in your pipeline from `base_project` in `new_project`.

You can also interactively run pipelines, or schedule them in jobs. If you interactively run a pipeline from `base_project`, you need to create a Config in `new_project`. Then, the pipeline is visible in `new_project`.

For jobs, you do not need to import a pipeline. When you create a job in `new_project`, you can select any pipeline from `base_project` in the pipeline operator. All Config values from `base_project` and `new_project` are then available here in the job.

> If you modify anything from an existing or new pipeline, Prophecy updates `new_project` after you release `base_project`, and updates the dependency in `new_project`.

Author: (For Monday) Question: Would rephrasing line 39 to the following capture what you're saying:

```suggestion
If you modify anything from an existing or new pipeline, you must publish these updates as a new version for the package to pick up the changes. Then, you must manually update the project dependency to get the updated package version.
```

> If you modify anything from an existing or new pipeline, Prophecy updates `new_project` after you release `base_project`, and updates the dependency in `new_project`.

Reviewer: Let's see:

> If you modify anything from an existing or new pipeline

I think that this is wrong phrasing. It's more like "If you modify the original pipeline of the packaged project"

> you must publish these updates as a new version for the package to pick up the changes

you must release a new version of the project: https://docs.prophecy.io/engineers/deployment

this creates a new project version, which corresponds to a new package version.

> Then, you must manually update the project dependency to get the updated package version.

This is right.
